Split-Sending

MrTJP opened the pull request:
-Basic item splitsending between chassis.
-each pipe in a group will be served
-"amount" parameter specifies how many items to give that pipe before
moving on to the next one
-functional, but could definately use some cleanup.

Commit:
Split-Sending

-Basic item splitsending between chassis.
-each pipe in a group will be served
-"amount" parameter specifies how many items to give that pipe before
moving on to the next one
-functional, but could definately use some cleanup.

AartBluestoke commented on common/logisticspipes/pipes/PipeLogisticsChassi.java:
```diff
@@ -280,28 +285,32 @@ public void sendStack(ItemStack stack, int destination, ItemSendMode mode, List<
 				convertFromMeta = true;
 			}
 			switchOrientationOnTick = (ChassiLogic.orientation == ForgeDirection.UNKNOWN);
+			splitGroup = nbt.getInteger("SplitGroup");
```
how will this handle updating from old values, where this int is not present in the nbt? (i'm not saying this is an error, i'm just not sure what happens when the token isn't there, and we don't want to throw errors as that'll nuke every single chassi.

MrTJP replied:
nothing would happen.  it would just use the default values of 0 group and 1 split, and it will just be created next time it saves the NBT.  i havent seen any errors from this.

AartBluestoke commented on common/logisticspipes/logistics/LogisticsManagerV2.java:
```diff
@@ -184,9 +188,105 @@ public static SinkReply canSink(IRouter destination, IRouter sourceRouter, boole
 	}
 	
+	//UUID, group, member dataq
+	public List<Pair3<UUID, Integer, SplitMember>> groups = new LinkedList<Pair3<UUID, Integer, SplitMember>>();
```
make this private or protected; we don't want it being manipulated except through the visible functions.

Also, this should me a map<Integer,SplitMember>, rather than that list - that will make all the for loops below collapse to a groups.get(key); the code will also be more readable if the UUID is moved to the splitMember, so that you don't need to carry around the pairs of UUID,SplitMember.

Finally, for speed of code, routers can use a simpleID, rather than a UUID (even though that doesn't persist between startup), as all that needs to be persisted is the groupID. (handling 2 longs rather than an int can make a difference, especially if you move to a hashmap, where the UUID is being hashed).

MrTJP replied:
!!!
I thought that each item in a Map needed a unique key? So if they are in the same group, wouldnt they need the same key? unless i hashed it as ID, SplitMember instead of GroupID, SplitMember.

AartBluestoke commented on common/logisticspipes/utils/SinkReply.java:
```diff
@@ -18,7 +18,7 @@
 	public final boolean isPassive;
 	public final boolean isDefault;
 	public final int energyUse;
-	public final int maxNumberOfItems;
```
this should stay as final unless you've got a good reason (this patch doesn't contain any changes to it); a) for java compiled code optimisation, and b) if this can change, then we _must_ duplicate it before any possible change, as i think several parts of the code assume that it doesn't get changed

MrTJP replied:
Sorry, ur right. I intended to change this back to how it was. 

On Apr 28, 2013, at 2:48 AM, Andrew Hill notifications@github.com wrote:

> In common/logisticspipes/utils/SinkReply.java:
> 
> > @@ -18,7 +18,7 @@
> >     public final boolean isPassive;
> >     public final boolean isDefault;
> >     public final int energyUse;
> > -   public final int maxNumberOfItems;
> >   this should stay as final unless you've got a good reason (this patch doesn't contain any changes to it); a) for java compiled code optimisation, and b) if this can change, then we must duplicate it before any possible change, as i think several parts of the code assume that it doesn't get changed
> 
> —
> Reply to this email directly or view it on GitHub.

AartBluestoke commented:
quick question: what is the intended use case of this?

MrTJP commented:
Intended use:
For example you have all of your honey coming in from bees and you want half of it to be put in a tank, and the other half to go into a chest. Basically, it will say that these two chassi pipes are on the same priority and cost. The items will be evenly sinced between them. 

But like I said, it is horrible code, it has to be cleaned. I was just tired so I published it and went to bed. 

On Apr 28, 2013, at 2:53 AM, Andrew Hill notifications@github.com wrote:

> quick question: what is the intended use case of this?
> 
> —
> Reply to this email directly or view it on GitHub.

MrTJP commented:
problem: if a pipe is configured to share, but it can no longer accept items, they wont pass the turn to someone else.    Because the pipe that wont accept items anymore isnt sending sinc replies, the turn wont be passed.  In which case, items will just make their way over to another route.

Commit:
Split sending cleanup

-Used simple ID to improve performance.
-Bugs are gone now, its working as intended.
-Basically, 2 linked chassis are treated as 1 inventory.
-If one fills up, neither will accept items.
-Performance impact is nothing to worry about, especially since only a
few pipes in a world will ever use this feature anyway.

davboecki commented on common/logisticspipes/pipes/basic/CoreRoutedPipe.java:
```diff
@@ -304,7 +302,6 @@ public final void updateEntity() {
 				}
 				//first tick just create a router and do nothing.
 				getRouter();
-				return;
```
Don't remove that.

davboecki commented on common/logisticspipes/pipes/basic/CoreRoutedPipe.java:
```diff
@@ -290,7 +288,7 @@ public void enabledUpdateEntity() {}
 	public void ignoreDisableUpdateEntity() {}
 	
 	@Override
-	public final void updateEntity() {
```
Don't remove this.

MrTJP commented:
this needs some reworking....
